cf-file-scanner/src/main/java/org/broadinstitute/ddp/cf/FileScanner.java

```diff
@@ -0,0 +1,280 @@
+                "--config-file=" + FRESHCLAM_CONF,
+                "--datadir=" + DB_DIR);
+        logger.info("Output: " + result.getOutput());
+        logger.info("Status: " + result.getExitCode());
+
+        int status = result.getExitCode();
+        if (status == 0) {
+            dbUpdatedAt = Instant.now();
+            long elapsed = System.currentTimeMillis() - start;
+            logger.info("Finished refreshing the database at: " + dbUpdatedAt + " (took " + elapsed + "ms)");
+        } else {
+            throw new RuntimeException("Unable to refresh database, freshclam exited with status: " + status);
+        }
+    }
+
+    private static ScanResult runClamscan(byte[] content) {
+        CommandResult result = runCommand(true, content,
+                CLAMSCAN_BIN,
+                "--database=" + DB_DIR,
+                "-");   // Dash means to read from stdin.
+        logger.info("Output: " + result.getOutput());
+        logger.info("Status: " + result.getExitCode());
+
+        int status = result.getExitCode();
+        if (status == 0) {
+            return ScanResult.CLEAN;
+        } else if (status == 1) {
+            return ScanResult.INFECTED;
+        } else {
+            throw new RuntimeException("Error while scanning file, clamscan exited with status: " + status);
+        }
+    }
+
+    @Override
+    public void accept(Message message, Context context) {
+        logger.info("Received message: eventId=" + context.eventId() + " timestamp=" + context.timestamp());
+
+        String bucketName = message.getAttributes().get(ATTR_BUCKET_ID);
+        String fileName = message.getAttributes().get(ATTR_OBJECT_ID);
+        if (bucketName == null || bucketName.isBlank()) {
+            logger.severe("Bucket name is missing in message");
+            return;
+        } else if (fileName == null || fileName.isBlank()) {
+            logger.severe("Bucket filename is missing in message");
+            return;
+        }
+
+        BlobId blobId = BlobId.of(bucketName, fileName);
+        logger.info("Looking up file: " + blobId.toString());
+        Blob blob = storage.get(blobId);
+        if (blob == null || !blob.exists()) {
+            logger.severe("Could not find file: " + blobId.toString());
+            // This might be a duplicate (due to pubsub at-least-once delivery), in which case the
+            // file might have already been moved. Exit gracefully so we don't incur a cold start.
+            return;
+        }
+
+        // Ensure database is up-to-date before scanning.
+        checkDatabaseFiles();
+
+        logger.info("Scanning file: " + blobId.toString());
+        ScanResult result = runClamscan(blob.getContent());
+
+        String data = message.getData() != null ? message.getData() : "";
+        var resultMessage = PubsubMessage.newBuilder()
+                .putAllAttributes(message.getAttributes())
+                .putAttributes(ATTR_SCAN_RESULT, result.name())
```

Comment:
I'm curious if we should include a link to the Cloud Logging console message with the reason for failure (or the entire message as a data field or an attribute, if the space allows). Either way, this could help easily determine why a file didn't pass the scan downstream.

Reply:
Currently, I don't see us needing that information, but could be worked out in the future!